File: TinyScript/java/src/test/java/lexer/TokenTest.java
package lexer;

import commons.PeekIterator;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/7/7 0:09
 **/
class TokenTest {

    @Test
    public void testVarOrKeyword() {
        var it1 = new PeekIterator<>("if abc".chars().mapToObj(x -> (char) x));
        var it2 = new PeekIterator<>("true abc".chars().mapToObj(x -> (char) x));
        Token token1 = Token.makeVarOrKeyword(it1);

        Token token2 = Token.makeVarOrKeyword(it2);

        assertEquals(token1.getType(), TokenType.KEYWORD);
        assertEquals(token1.getValue(), "if");
        assertEquals(token2.getType(), TokenType.BOOLEAN);
        assertEquals(token2.getValue(), "true");


        var it3 = new PeekIterator<>("abc".chars().mapToObj(x -> (char) x));

        Token token3 = Token.makeVarOrKeyword(it3);
        assertEquals(token3.getType(), TokenType.VARIABLE);
        assertEquals(token3.getValue(), "abc");

    }

    @Test
    public void makeString() throws LexicalException {
        String[] tests = {
                "\"123\"",
                "'123'",
        };

        List<PeekIterator<Character>> collect = Stream.of(tests).map((item -> new PeekIterator<>(item.chars().mapToObj(x -> (char) x)))).collect(Collectors.toList());
        for (PeekIterator<Character> characterPeekIterator : collect) {
            Token token = Token.makeString(characterPeekIterator);
            Assertions.assertEquals(token.getValue(), "123");
        }
    }

    @Test
    public void testMakeOp() throws LexicalException {
        String[] tests = {
                "+ xx",
                "++mm",
                "- xx",
                "--mm",
                "/=g",
                "= 1",
                "==sad",
                "&=sada",
                "&yy",
                "||xxs",
                "^=11",
                "%7"

        };
        int i = 0;
        String[] results = {"+", "++", "-", "--", "/=", "=", "==", "&=", "&", "||", "^=", "%"};
        for (String test : tests) {
            PeekIterator<Character> iterator = new PeekIterator<>(test.chars().mapToObj(x -> (char) x), (char) 0);
            Token token = Token.makeOp(iterator);
            Assertions.assertEquals(token.getValue(), results[i++]);
            Assertions.assertEquals(token.getType(), TokenType.OPERATOR);
        }
    }

    @Test
    void testMakeNumber() throws LexicalException {
        String[] tests = {
                "0000xx",
                "0.21d1",
                "021dew",
                "2.11sd",
                "21sa",
                "-1xx",
                "+1a",
                "1231.1231x"
        };
        String[] results = {"0", "0.21", "21", "2.11", "21", "-1", "+1", "1231.1231"};
        int i = 0;
        for (String test : tests) {
            PeekIterator<Character> iterator = new PeekIterator<>(test.chars().mapToObj(x -> (char) x), (char) 0);
            Token token = Token.makeNumber(iterator);
            Assertions.assertEquals(token.getValue(), results[i++]);

        }
    }
}

File: TinyScript/js/src/test/commons/Token.test.ts
import PeekIterator from "../../main/js/commons/PeekIterator";
import arrayToGenerator from "../../main/js/commons/ArrayToGenerator";
import Token from "../../main/js/lexer/Token";
import TokenType from "../../main/js/lexer/TokenType";
import * as assert from "assert";
import ASTNodeTypes from "../../main/js/parser/ast/ASTNodeTypes";

function assertToken(token:Token, value: string, type: TokenType) {
    assert.equal(token.value, value)
    assert.equal(token.type, type)
}

describe("test token", () => {
    test('test VarOrKeyword', () => {
        const data = 'true asdas'
        const it: PeekIterator<string> = new PeekIterator<string>(arrayToGenerator<string>([...data]))
        const token: Token = Token.makeVarOrKeyword(it);
        expect(token.value).toEqual("true")
        expect(token.type).toEqual(TokenType.BOOLEAN)

        const data1 = 'if asdas'
        const it1: PeekIterator<string> = new PeekIterator<string>(arrayToGenerator<string>([...data1]))
        const token1: Token = Token.makeVarOrKeyword(it1);
        expect(token1.value).toEqual("if")
        expect(token1.type).toEqual(TokenType.KEYWORD)
    })
    test('test makeString', () => {
        const tests: string[] = [
            "'123'",
            '"123"',
        ];
        const peekIterators: PeekIterator<string>[] = tests.map(item => {
            return new PeekIterator<string>(arrayToGenerator([...item]))
        });
        for (const [index, peekIterator] of peekIterators.entries()) {
            const token2 = Token.makeString(peekIterator);
            assertToken(token2, tests[index], TokenType.STRING);
        }
    })

    test("test make op", () => {
        const tests: string[] = [
            "+ xx",
            "++mm",
            "- xx",
            "--mm",
            "/=g",
            "= 1",
            "==sad",
            "&=sada",
            "&yy",
            "||xxs",
            "^=11",
            "%7"
        ];
        let i = 0;
        const results: string[] = ["+", "++", "-", "--", "/=", "=", "==", "&=", "&", "||", "^=", "%"];
        for (const test of tests) {
            const iterator: PeekIterator<string> = new PeekIterator<string>(arrayToGenerator([...test]));
            const token: Token = Token.makeOp(iterator);
            expect(token.value).toEqual(results[i++]);
            expect(token.type).toEqual(TokenType.OPERATOR);
        }
    })

    test("test make number", () => {
        const tests: string[] = [
            "10",
            "0000xx",
            "0.21d1",
            "021dew",
            "2.11sd",
            "21sa",
            "-1xx",
            "20",
            "1231.1231x"
        ];
        const results: string[] = ["10", "0", "0.21", "21", "2.11", "21", "-1", "20", "1231.1231"];
        let i = 0;
        for (const test of tests) {
            const iterator: PeekIterator<string> = new PeekIterator<string>(arrayToGenerator([...test]), '\0');
            const token3 = Token.makeNumber(iterator);
            expect(token3.value).toEqual(results[i++]);
        }
    })
})